log4j-core/src/main/java/org/apache/logging/log4j/core/util/DefaultShutdownCallbackRegistry.java

```diff
@@ -87,42 +92,41 @@ public void run() {
     }
 
     private static class RegisteredCancellable implements Cancellable {
-        // use a reference to prevent memory leaks
-        private final Reference<Runnable> hook;
-        private Collection<Cancellable> registered;
+        private Runnable callback;
+        private Collection<Reference<Cancellable>> registered;
 
-        RegisteredCancellable(final Runnable callback, final Collection<Cancellable> registered) {
+        RegisteredCancellable(final Runnable callback, final Collection<Reference<Cancellable>> registered) {
+            this.callback = callback;
             this.registered = registered;
-            hook = new SoftReference<>(callback);
         }
 
         @Override
         public void cancel() {
-            hook.clear();
-            registered.remove(this);
+            callback = null;
+            registered.removeIf(ref -> ref.get() == this);
```

Comment:
Might as well include a null check as well to remove reference objects whose target has been garbage collected while we're at it.

Reply:
I'm going to update this to include the null check to remove collected values, and add a null check on registered so that subsequent `cancel()` invocations do not NPE.